### DISCUSSION TITLE
Bug in developer documentation

### DISCUSSION THREAD

**Discussion by worldpeace-germany** [missionpinball/missionpinball#140] [Q&A]
Hello,

I am referring to

https://developer.missionpinball.org/en/dev/code/machine_code.html

unfortunately the docs are generated from code comments as far as I know and I am not too sure how to edit them best. Maybe either someone can give me a hint or let me know if there is a newer documentation or do the change in the code.

It says

>  First, add a code folder to your machine folder (you can use another name if you want).

  Then the screenshot below shows a folder scriplets. If I place my code in the scriplets folder I am getting it to work, I am not able to get it to work in a folder named code. Maybe that can get some clarifications.

In addition I believe the `__init__.py` is not needed anymore these days.

Best,
Joe

**Comment by avanwinkle** [comment]
That specific link is for the 0.54 documentation, which is quite outdated. It looks like that portion of the documentation didn't get migrated during the big documentation refactor/rewrite last year, but the short answer is that your custom code can _technically_ live in any folder because you specify the folder as part of the module name. The (new) expected default is "custom_code", so your config would look like this:

```
custom_code:
  - custom_code.leaderboard_tracker.LeaderboardTracker
  - custom_code.some_other_filename.CustomCodeClassToImport
```

All of the documentation comes from the [MPF Docs repository](https://github.com/missionpinball/mpf-docs), and all improvements and contributions are welcome!

**Comment by worldpeace-germany** [comment]
I am just behind you on the contributer list for this mpf-docs repro, trying to catch up ;-) But honestly, for the developer documentation I tried google, I tried github search, I tried search on missionpinball.org All I find is the old documentation. Your link is to the general mpf doc repro, but not the developer pages. Do you have link pointing to that one directly?

Yes it was probably not migrated because at that time the developer docs was not part of mpf-docs but was generated out of documentation which lived inside the code as code comments. So if there is non at all, then we could think of what would be best and easiest to get it somehow into the mpf-docs. Happy to contribute, but we probably should align on a way forward.

To the technical issue:
- servlet as subfolder works
- custom_code as subfolder works
- any other folder I am not able to get to work - I can live with the default, so no urgent matter for me just in case we want to have it documented for others we should get it to work

**Comment by worldpeace-germany** [comment]
@avanwinkle I have now created a pull request where I have included all the old developer documentation to be part of the mpf-docs repro. I hope it was time well spent and the pull request can be accepted. It is basically 98% the original documentation, I made some minor adjustments:
- added a debugging chapter
- removed some empty pages and clubbed a few very short pages together
- ignored the bcp API since my understanding is that this will go away with 0.80 so it is not needed anymore.

If there is any issue with my changes please let me know, I am happy to improve where necessary, just reach out.

**Comment by avanwinkle** [comment]
Thanks so much! Left some notes on the PR, you're amazing for taking on this task!

FYI I've started working on extending the MKDocs implementation that hosts the documentation to allow version-splitting, so we can have one branch for the 0.57 documentation and a separate branch for the 0.80 documentation. This will allow us to start culling the MPF-MC config references and making the GMC documentation more prominent without causing so much confusion.

I was not entirely successful yesterday and the docs were offline for a bit. They're back up now but the versioning isn't working yet, I'll keep at it this week.

**Comment by avanwinkle** [comment]
Pull request merged, thanks so much!

The BCP interface is still applicable to MPF 0.80, that documentation is primarily for people who don't use a built-in Media Controller (MC for 0.57, GMC for 0.80) and want to build their own. In fact, I was building my own media controller in Godot using the BCP interface before I turned my work into GMC.

I've also successfully got the versioning working on the docs, so the default is now https://missionpinball.org/latest/ (aka https://missionpinball.org/0.57/) and a separate branch of the docs repo will publish to https://missionpinball.org/0.80/. I'd like to add a banner to the 0.80 docs so it's clear when folks are there, and I haven't yet figured out the version dropdown selector. But it's getting there!

**Comment by worldpeace-germany** [comment]
Will add the missing BCP interfaces in the near future, shouldn't be too much. 

First I need to understand the concept/idea of the branches.We have now main, 0.57, 0.80. To what branch should I commit? Always main and you decide if it is for 0.57 or 0.80? Or should I commit to 0.57/0.80 if version dependent and to main if for both? For 0.80 for example I would like to remove one page from the newly created developer docs (which btw is not yet merged into 0.80), but if later someone changes something in main on that page of course it should not show up in 0.80 anymore as a "new" page. How is that handled?

I would add that info as well to the beginners guide for docs update - btw the link "Something missing or wrong? You can fix it!" points to  https://missionpinball.org/about/help and throws an 404 since the version number is missing in the URI.

PS: making good progress on the segment_display, being pretty confident that I will get it to work.
